stat/stat_test.go

```diff
@@ -324,78 +259,13 @@ func TestCrossEntropy(t *testing.T) {
 	}
 }
 
-func ExampleEntropy() {
-
-	p := []float64{0.05, 0.1, 0.9, 0.05}
-	entP := Entropy(p)
-
-	q := []float64{0.2, 0.4, 0.25, 0.15}
-	entQ := Entropy(q)
-
-	r := []float64{0.2, 0, 0, 0.5, 0, 0.2, 0.1, 0, 0, 0}
-	entR := Entropy(r)
-
-	s := []float64{0, 0, 1, 0}
-	entS := Entropy(s)
-
-	fmt.Println("Entropy is a measure of the amount of uncertainty in a distribution")
-	fmt.Printf("The second bin of p is very likely to occur. It's entropy is %.4f\n", entP)
-	fmt.Printf("The distribution of q is more spread out. It's entropy is %.4f\n", entQ)
-	fmt.Println("Adding buckets with zero probability does not change the entropy.")
-	fmt.Printf("The entropy of r is: %.4f\n", entR)
-	fmt.Printf("A distribution with no uncertainty has entropy %.4f\n", entS)
-	// Output:
-	// Entropy is a measure of the amount of uncertainty in a distribution
-	// The second bin of p is very likely to occur. It's entropy is 0.6247
-	// The distribution of q is more spread out. It's entropy is 1.3195
-	// Adding buckets with zero probability does not change the entropy.
-	// The entropy of r is: 1.2206
-	// A distribution with no uncertainty has entropy 0.0000
-}
-
-func ExampleExKurtosis() {
-	fmt.Println(`Kurtosis is a measure of the 'peakedness' of a distribution, and the
-excess kurtosis is the kurtosis above or below that of the standard normal
-distribution`)
-	x := []float64{5, 4, -3, -2}
-	kurt := ExKurtosis(x, nil)
-	fmt.Printf("ExKurtosis = %.5f\n", kurt)
-	weights := []float64{1, 2, 3, 5}
-	wKurt := ExKurtosis(x, weights)
-	fmt.Printf("Weighted ExKurtosis is %.4f", wKurt)
-	// Output:
-	// Kurtosis is a measure of the 'peakedness' of a distribution, and the
-	// excess kurtosis is the kurtosis above or below that of the standard normal
-	// distribution
-	// ExKurtosis = -5.41200
-	// Weighted ExKurtosis is -0.6779
-}
-
 func TestExKurtosis(t *testing.T) {
 	// the example does a good job, this just has to cover the panic
 	if !panics(func() { ExKurtosis(make([]float64, 3), make([]float64, 2)) }) {
 		t.Errorf("ExKurtosis did not panic with x, weights length mismatch")
 	}
 }
```

Comment:
For example this.

Reply:
well, this just means we rely on `ExampleExKurtosis` to do a good job at testing `ExKurtosis`.
the comment in `TestExKurtosis` could be modified to explicitly mention `ExampleExKurtosis` instead of just "the example...".

examples are for users.
tests are for developers.

sometimes they (users/devs) overlap.
but the most important thing (for users and for devs) is how the API looks like and how it should be used.
putting examples into their own file isn't the main point: it's to put them into a file with a `package xxx_test` clause.

Reply:
Yeah. OK.

Reply:
you mean migrate `stat_test.go` from `package stat` to `package stat_test` ?

Reply:
Yeah. Something like this; it's a minimal change (all the tests could be moved into `stat_test`, but it's not necessary).

```diff
diff --git a/stat/faithful_test.go b/stat/faithful_test.go
index 77df6e2b..0028b7ce 100644
--- a/stat/faithful_test.go
+++ b/stat/faithful_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-package stat
+package stat_test
 
 // faithful is the faithful data set from R.
 var faithful = struct{ waiting, eruptions []float64 }{
diff --git a/stat/stat_test.go b/stat/stat_test.go
index 995e0b0e..8cfb21b8 100644
--- a/stat/stat_test.go
+++ b/stat/stat_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-package stat
+package stat_test
 
 import (
        "fmt"
@@ -14,12 +14,14 @@ import (
        "golang.org/x/exp/rand"
 
        "gonum.org/v1/gonum/floats"
+       "gonum.org/v1/gonum/stat"   // Named import for examples.
+       . "gonum.org/v1/gonum/stat" // Dot import for tests.
 )
 
 func ExampleCircularMean() {
        x := []float64{0, 0.25 * math.Pi, 0.75 * math.Pi}
        weights := []float64{1, 2, 2.5}
-       cmean := CircularMean(x, weights)
+       cmean := stat.CircularMean(x, weights)
 
        fmt.Printf("The circular mean is %.5f.\n", cmean)
        // Output:
@@ -70,7 +72,7 @@ func ExampleCorrelation() {
        fmt.Println("Correlation computes the degree to which two datasets move together")
        fmt.Println("about their mean. For example, x and y above move similarly.")
 
-       c := Correlation(x, y, w)
+       c := stat.Correlation(x, y, w)
        fmt.Printf("Correlation is %.5f\n", c)
 
        // Output:
@@ -147,7 +149,7 @@ func ExampleKendall() {
        fmt.Println("Kendall correlation computes the number of ordered pairs")
        fmt.Println("between two datasets.")
 
-       c := Kendall(x, y, w)
+       c := stat.Kendall(x, y, w)
        fmt.Printf("Kendall correlation is %.5f\n", c)
 
        // Output:
@@ -221,12 +223,12 @@ func ExampleCovariance() {
        fmt.Println("about their mean.")
        x := []float64{8, -3, 7, 8, -4}
        y := []float64{10, 2, 2, 4, 1}
-       cov := Covariance(x, y, nil)
+       cov := stat.Covariance(x, y, nil)
        fmt.Printf("Cov = %.4f\n", cov)
        fmt.Println("If datasets move perfectly together, the variance equals the covariance")
        y2 := []float64{12, 1, 11, 12, 0}
-       cov2 := Covariance(x, y2, nil)
-       varX := Variance(x, nil)
+       cov2 := stat.Covariance(x, y2, nil)
+       varX := stat.Variance(x, nil)
        fmt.Printf("Cov2 is %.4f, VarX is %.4f", cov2, varX)
        // Output:
        // Covariance computes the degree to which datasets move together
@@ -325,18 +327,17 @@ func TestCrossEntropy(t *testing.T) {
 }
 
 func ExampleEntropy() {
-
        p := []float64{0.05, 0.1, 0.9, 0.05}
-       entP := Entropy(p)
+       entP := stat.Entropy(p)
 
        q := []float64{0.2, 0.4, 0.25, 0.15}
-       entQ := Entropy(q)
+       entQ := stat.Entropy(q)
 
        r := []float64{0.2, 0, 0, 0.5, 0, 0.2, 0.1, 0, 0, 0}
-       entR := Entropy(r)
+       entR := stat.Entropy(r)
 
        s := []float64{0, 0, 1, 0}
-       entS := Entropy(s)
+       entS := stat.Entropy(s)
 
        fmt.Println("Entropy is a measure of the amount of uncertainty in a distribution")
        fmt.Printf("The second bin of p is very likely to occur. It's entropy is %.4f\n", entP)
@@ -358,10 +359,10 @@ func ExampleExKurtosis() {
 excess kurtosis is the kurtosis above or below that of the standard normal
 distribution`)
        x := []float64{5, 4, -3, -2}
-       kurt := ExKurtosis(x, nil)
+       kurt := stat.ExKurtosis(x, nil)
        fmt.Printf("ExKurtosis = %.5f\n", kurt)
        weights := []float64{1, 2, 3, 5}
-       wKurt := ExKurtosis(x, weights)
+       wKurt := stat.ExKurtosis(x, weights)
        fmt.Printf("Weighted ExKurtosis is %.4f", wKurt)
        // Output:
        // Kurtosis is a measure of the 'peakedness' of a distribution, and the
@@ -382,7 +383,7 @@ func ExampleGeometricMean() {
        x := []float64{8, 2, 9, 15, 4}
        weights := []float64{2, 2, 6, 7, 1}
        mean := Mean(x, weights)
-       gmean := GeometricMean(x, weights)
+       gmean := stat.GeometricMean(x, weights)
 
        logx := make([]float64, len(x))
        for i, v := range x {
@@ -425,8 +426,8 @@ func TestGeometricMean(t *testing.T) {
 func ExampleHarmonicMean() {
        x := []float64{8, 2, 9, 15, 4}
        weights := []float64{2, 2, 6, 7, 1}
-       mean := Mean(x, weights)
-       hmean := HarmonicMean(x, weights)
+       mean := stat.Mean(x, weights)
+       hmean := stat.HarmonicMean(x, weights)
 
        fmt.Printf("The arithmetic mean is %.5f, but the harmonic mean is %.4f.\n", mean, hmean)
        // Output:
@@ -579,7 +580,7 @@ and 0 data points above 1000. Since dividers has length 5, there will be 4 bins.
        max++
        floats.Span(dividers, min, max)
        // Span includes the min and the max. Trim the dividers to create 10 buckets
-       hist = Histogram(nil, dividers, x, nil)
+       hist = stat.Histogram(nil, dividers, x, nil)
        fmt.Printf("Hist = %v\n", hist)
        fmt.Println()
        fmt.Println(`Histogram also works with weighted data, and allows reusing of
@@ -588,7 +589,7 @@ the count field in order to avoid extra garbage`)
        for i := range weights {
                weights[i] = float64(i + 1)
        }
-       Histogram(hist, dividers, x, weights)
+       stat.Histogram(hist, dividers, x, weights)
        fmt.Printf("Weighted Hist = %v\n", hist)
 
        // Output:
@@ -858,14 +859,13 @@ func TestKolmogorovSmirnov(t *testing.T) {
 }
 
 func ExampleKullbackLeibler() {
-
        p := []float64{0.05, 0.1, 0.9, 0.05}
        q := []float64{0.2, 0.4, 0.25, 0.15}
        s := []float64{0, 0, 1, 0}
 
-       klPQ := KullbackLeibler(p, q)
-       klPS := KullbackLeibler(p, s)
-       klPP := KullbackLeibler(p, p)
+       klPQ := stat.KullbackLeibler(p, q)
+       klPS := stat.KullbackLeibler(p, s)
+       klPP := stat.KullbackLeibler(p, p)
 
        fmt.Println("Kullback-Leibler is one measure of the difference between two distributions")
        fmt.Printf("The K-L distance between p and q is %.4f\n", klPQ)
@@ -1163,13 +1163,13 @@ func TestHellinger(t *testing.T) {
 
 func ExampleMean() {
        x := []float64{8.2, -6, 5, 7}
-       mean := Mean(x, nil)
+       mean := stat.Mean(x, nil)
        fmt.Printf("The mean of the samples is %.4f\n", mean)
        w := []float64{2, 6, 3, 5}
-       weightedMean := Mean(x, w)
+       weightedMean := stat.Mean(x, w)
        fmt.Printf("The weighted mean of the samples is %.4f\n", weightedMean)
        x2 := []float64{8.2, 8.2, -6, -6, -6, -6, -6, -6, 5, 5, 5, 7, 7, 7, 7, 7}
-       mean2 := Mean(x2, nil)
+       mean2 := stat.Mean(x2, nil)
        fmt.Printf("The mean of x2 is %.4f\n", mean2)
        fmt.Println("The weights act as if there were more samples of that number")
        // Output:
@@ -1544,11 +1544,11 @@ func TestQuantileInvalidCumulantKind(t *testing.T) {
 
 func ExampleStdDev() {
        x := []float64{8, 2, -9, 15, 4}
-       stdev := StdDev(x, nil)
+       stdev := stat.StdDev(x, nil)
        fmt.Printf("The standard deviation of the samples is %.4f\n", stdev)
 
        weights := []float64{2, 2, 6, 7, 1}
-       weightedStdev := StdDev(x, weights)
+       weightedStdev := stat.StdDev(x, weights)
        fmt.Printf("The weighted standard deviation of the samples is %.4f\n", weightedStdev)
        // Output:
        // The standard deviation of the samples is 8.8034
@@ -1558,10 +1558,10 @@ func ExampleStdDev() {
 func ExampleStdErr() {
        x := []float64{8, 2, -9, 15, 4}
        weights := []float64{2, 2, 6, 7, 1}
-       mean := Mean(x, weights)
-       stdev := StdDev(x, weights)
+       mean := stat.Mean(x, weights)
+       stdev := stat.StdDev(x, weights)
        nSamples := floats.Sum(weights)
-       stdErr := StdErr(stdev, nSamples)
+       stdErr := stat.StdErr(stdev, nSamples)
        fmt.Printf("The standard deviation is %.4f and there are %g samples, so the mean\nis likely %.4f ± %.4f.", stdev, nSamples, mean, stdErr)
        // Output:
        // The standard deviation is 10.5733 and there are 18 samples, so the mean
@@ -1735,11 +1735,11 @@ func TestVariance(t *testing.T) {
 
 func ExampleVariance() {
        x := []float64{8, 2, -9, 15, 4}
-       variance := Variance(x, nil)
+       variance := stat.Variance(x, nil)
        fmt.Printf("The variance of the samples is %.4f\n", variance)
 
        weights := []float64{2, 2, 6, 7, 1}
-       weightedVariance := Variance(x, weights)
+       weightedVariance := stat.Variance(x, weights)
        fmt.Printf("The weighted variance of the samples is %.4f\n", weightedVariance)
        // Output:
        // The variance of the samples is 77.5000
diff --git a/stat/statmat_test.go b/stat/statmat_test.go
index 8dfc72c5..424c43b2 100644
--- a/stat/statmat_test.go
+++ b/stat/statmat_test.go
@@ -296,6 +296,18 @@ func TestMahalanobis(t *testing.T) {
        }
 }
 
+// panics returns true if the called function panics during evaluation.
+func panics(fun func()) (b bool) {
+       defer func() {
+               err := recover()
+               if err != nil {
+                       b = true
+               }
+       }()
+       fun()
+       return
+}
+
 // benchmarks
 
 func randMat(r, c int) mat.Matrix {
```

Note that the `panics` helper needs duplication (or it could be exported in the tests in "statmat_test.go").